### PR TITLE
Add Filename.temp_dir function to create a temporary directory

### PR DESCRIPTION
**Pull request by novalis**
We recently wanted this for [Tezos](https://gitlab.com/tezos/tezos/-/merge_requests/4320) and didn't find it in the standard library.  Prior art: POSIX has mkdtemp.

### CODE REVIEW

File: stdlib/filename.ml
Comment (avsm) [inline stdlib/filename.ml]: The function looks good to me, but why did you pick 1000 retries? That seems like an awfully high number to spin on for avoiding a race: you are either in the presence of a malicious attacker at this point, or the server is under so much load that spinning won't help anyway. I notice that [musl uses a 100](https://github.com/bminor/musl/blob/master/src/temp/mkdtemp.c); I've not checked other mkdtemp implementations.

Comment (novalis) [reply stdlib/filename.ml]: Copy-paste from other examples in the same file.  Want me to submit a second PR to change them all?

File: testsuite/tests/lib-filename/temp.ml
Comment (v-gb) [inline testsuite/tests/lib-filename/temp.ml]: This test is reasonable, but has a number of issues:
- strange indentation
- `else ()` is fairly unusual
- cleans up tmp dirs on error but not on success
- failures in this test would contain no information, because it swalows errors

I'd recommend writing this as:
```ocaml
 let tmpdir1 = ref "" in
 let tmpdir2 = ref "" in
 Fun.protect ~finally:(fun () ->
   if !tmpdir1 <> "" then Sys.rmdir !tmpdir1;
   if !tmpdir2 <> "" then Sys.rmdir !tmpdir2;
 ) (fun () ->
    tmpdir1 := Filename.temp_dir "morx" "fleem";
    assert (Sys.is_directory !tmpdir1);
    tmpdir2 := Filename.temp_dir "morx" "fleem" ;
    assert (Sys.is_directory !tmpdir2);
    assert (!tmpdir1 <> !tmpdir2);
  )
```

File: stdlib/filename.ml
Comment (v-gb) [reply stdlib/filename.ml]: glibc uses [200k](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/posix/tempname.c;h=60f8541085091934230e0adf96ce8a241531ca18;hb=c923cd8c496c7f253f327361a65c737233c7ebbd#l247).

Given that 1000 is already used in other functions, and that it's vaguely the geometric mean of the retries in musl and glibc, it seems good to keep as is.

(FWIW, the retries are there to deal with collisions, not races. If some process is leaking temp files, it creates potential collisions but there's no race)

**Review by v-gb — CHANGES_REQUESTED** [review]
This function clearly fits in the stdlib, and the implementation looks good, modulo comment. I'm surprised that the stdlib is not already providing this, actually.

I initially thought this was prone to attacks where someone creates a symlink for the directory you're about to create, because there's no O_EXCL in mkdir. But it's not the case. And you can see musl and glbc implement this the same way (except they only retry on EEXIST).

This needs to be rebased though, as the rest of the `Filename` module has changed to be multicore-compatible. The Changes entry will also need to be moved to the latest version.

File: Changes
Comment (v-gb) [inline Changes]: This is in the 5.0 section, but 5.0 is already released. You should put this in the "Working version" section (you may need to create a "standard library" subsection if if doesn't exist yet).

File: stdlib/filename.mli
Comment (v-gb) [inline stdlib/filename.mli]: Something I forgot to say earlier: you should add a `@since 5.1` tag to the doc comment. Also, it'd be better to put this after `open_temp_file` to avoid separating `temp_file` and `open_temp_file`.

**Review by v-gb — CHANGES_REQUESTED** [review]

**Review by nojb — DISMISSED** [review]
LGTM

Please:

- Add `@since 5.1` annotation to the docstring,
- Move `Changes` entry to the right place, and add the list of reviewers (on its own line, like the rest of the entries)
- Remove the trailing empty line the test (to satisfy `check-typo`)

As this touches the standard library, a second official approval is needed.

File: Changes
Comment (v-gb) [inline Changes]: Leftover empty line here

```suggestion

```

**Review by v-gb — APPROVED** [review]

File: testsuite/tests/lib-filename/temp.ml
Comment (nojb) [inline testsuite/tests/lib-filename/temp.ml]: Please remove this empty line.

File: Changes
Comment (nojb) [inline Changes]: Please revert this unrelated change.

File: Changes
Comment (nojb) [inline Changes]: Please move the author to its own line, and add the list of reviewers.

File: Changes
Comment (nojb) [reply Changes]: This line was removed by mistake. Can you restore it?

**Review by v-gb — APPROVED** [review]

File: stdlib/filename.mli
Comment (dbuenzli) [inline stdlib/filename.mli]: I thought the idea was to rather tell people to use `Unix` than add more system stuff of that kind to the stdlib. Oh well, the `Filename` module is kind of a lost cause anyways.

In any case:

1. I think it would be good to add an optional argument `perms` to specifiy the permissions.
2. It would be good to significantly rework this doc string which is very confusing, 
   in particular you talk about two temporary directories regardless. I would start with: 
   ```
    [temp_dir ~perms ~temp_dir prefix suffix] creates and returns a fresh directory with 
    permissions [perms] (defaults to 0o700) inside [temp_dir]. …`.
   ```
3. Mention that the function fails if [temp_dir] does not exist (i.e. this does not create the path).

File: stdlib/filename.ml
Comment (dbuenzli) [inline stdlib/filename.ml:374]: The retry logic is going to trigger if `temp_dir` does not exist. I don't think this is a good idea, check it for existence first.

**Review by dbuenzli — CHANGES_REQUESTED** [review]

File: stdlib/filename.mli
Comment (nojb) [reply stdlib/filename.mli]: > It would be good to significantly rework this doc string which is very confusing,

I agree, but then both the docstrings of `temp_file` and `temp_dir` should be reworked in parallel, as they are facsimiles of each other.

File: stdlib/filename.ml
Comment (nojb) [reply stdlib/filename.ml:374]: This is the same logic as in `temp_file`, so if doing this change it should be done in both functions. (I am not opposed to it.)

File: stdlib/filename.mli
Comment (dbuenzli) [reply stdlib/filename.mli]: Well let's first have a good one for this one.

File: stdlib/filename.ml
Comment (dbuenzli) [reply stdlib/filename.ml:374]: Well now that we are here let's avoid the future issue: "`Filename.temp_{file,dir}` takes a long time to fail if `temp_dir` does not exist".

**Comment by nojb** [comment]
@novalis: are you able to make the two changes that were suggested by @dbuenzli?

- fail fast if `temp_dir` does not exist
- reword the docstring along the lines of https://github.com/ocaml/ocaml/pull/10967#discussion_r1059878326

(the changes should be done in both `Filename.temp_file` and `Filename.temp_dir`.)

**Comment by novalis** [comment]
Sorry for the delay.

File: stdlib/filename.ml
Comment (v-gb) [reply stdlib/filename.ml:374]: I don't think we should be adding such a check. The temp directory not existing is only one possible way for the function to fail after iterating all the way, but running out of space or inodes is more likely and can't be checked in such a way.

A better behaved function would be better, but it's not a big deal to spend 2ms failing because the whole program is likely to crash as a result anyway.

Maybe the doc can be tweaked but I'm in favor of proceeding as is otherwise.

**Comment by novalis** [comment]
On reflection, I also don't think I want to check if the temporary directory exists. First, because it optimizes for an error case, and second, because in general this sort of check is a bad practice as it's prone to TOCTOU issues. In this case there is no such issue, but it has the appearance of impropriety.

But I am not going to change the PR back until I hear a consensus from the reviewers.

**Comment by dbuenzli** [comment]
> First, because it optimizes for an error case,

It's not only about failing fast, more importantly it's about reporting good, non misleading, error messages whenever you can. As far as TOCTOU is concerned, there's only so much you can do, the interface POSIX gives you to the file system is mostly racy beyond hope.

**Comment by dbuenzli** [comment]
Note that going back to 

> I thought the idea was to rather tell people to use Unix than add more system stuff of that kind to the stdlib.

one of the reasons is that the set of system functions from the stdlib prevents us from making good implementations for these kind of functions. With the precise error reports of `Unix` it allows you to make things less racy. For example in this case you would like to mkdir the temporary path directly and then retry only on `EEXIST`. See [here](https://github.com/b0-system/b0/blob/152c47864e9da478027bfa1113b892334abecffc/src/std/b0_std.ml#L2946-L2971) for example.

**Comment by novalis** [comment]
> > First, because it optimizes for an error case,
> 
> It's not only about failing fast, more importantly it's about reporting good, non misleading, error messages whenever you can. As far as TOCTOU is concerned, there's only so much you can do, the interface POSIX gives you to the file system is mostly racy beyond hope.

The error would (without this change) be ENOENT, which is sufficiently clear to me.

Is there a way to get to consensus for this?  I will not block consensus either way, but it seems like the reviewers have a disagreement.

**Comment by dbuenzli** [comment]
> The error would (without this change) be ENOENT, which is sufficiently clear to me.

The point is that by using `Unix` you error immediatetly without TOCTOU, not after 1000 retries, and you can report that the parent directory doesn't exist, which is the actual error, not that the temporary directory you are trying to create doesn't exist – which is confusing for end-users.

**Comment by nojb** [comment]
> Note that going back to
> 
> > I thought the idea was to rather tell people to use Unix than add more system stuff of that kind to the stdlib.
 
Adding a binding of [`mkdtemp(3)`](https://man7.org/linux/man-pages/man3/mkdtemp.3.html) to `Unix` makes sense, but:

- This function does not exist under Windows, so an implementation would have to be figured out;
- [`mkdtemp(3)`](https://man7.org/linux/man-pages/man3/mkdtemp.3.html) may not be available on some of the (UNIX-like) operating systems were OCaml is compiled (but this can be worked around with some `configure`-wizardry);
- [`mkdtemp(3)`](https://man7.org/linux/man-pages/man3/mkdtemp.3.html) (resp. its analog for files, [`mkstemp(3)`](https://man7.org/linux/man-pages/man3/mkstemp.3.html)) do not allow setting the "temporary directory" within which to create the directory (resp. file) which is quite useful in general (think an application that wants to keep their temporary files isolated in their own directory).

This (especially the last point) is why I think there is space to add this function to `Filename`. (But I'm happy to hear other opinions.)

> Is there a way to get to consensus for this? I will not block consensus either way, but it seems like the reviewers have a disagreement.

After sleeping on it, I am also thinking that, if we add this to `Filename`, then we have to accept that the error handling is suboptimal; and indeed handling the case where `temp_dir` does not exist is not worth it. The error message may not be as accurate as it could be, but as pointed out already in the discussion there are many other error conditions that may come up that we would not be detecting. If we want accurate error reporting, we need to do it in `Unix`.

**Comment by v-gb** [comment]
Every function that creates files in the stdlib, `Sys.mkdir`, `open_in_gen`, `open_out`, `Filename.temp_file` says "no such file or directory" when the directory of the requested file doesn't exist. I just don't think it makes sense to make one special case here.

**Comment by v-gb** [comment]
> * [`mkdtemp(3)`](https://man7.org/linux/man-pages/man3/mkdtemp.3.html) (resp. its analog for files, [`mkstemp(3)`](https://man7.org/linux/man-pages/man3/mkstemp.3.html)) do not allow setting the "temporary directory" within which to create the directory (resp. file) which is quite useful in general (think an application that wants to keep their temporary files isolated in their own directory).

I didn't try but I'd be very surprised if it wasn't meant to be used as `mkdtemp (tmpdir ^ "/" ^ prefix ^ "XXXXXX" ^ suffix)`.

A function in Unix would be obviously good, but I don't see why it should prevent making this small extension to `Filename`. It's natural to look for a way to make a temp dir next to the function that creates a temp file.

**Comment by nojb** [comment]
> I didn't try but I'd be very surprised if it wasn't meant to be used as `mkdtemp (tmpdir ^ "/" ^ prefix ^ "XXXXXX" ^ suffix)`.

Yes, you are right.

**Comment by dbuenzli** [comment]
> The error message may not be as accurate as it could be, but as pointed out already in the discussion there are many other error conditions that may come up that we would not be detecting. If we want accurate error reporting, we need to do it in `Unix`.

Well I think the other error conditions would be reported correctly or at least not in a misleading or puzzling way.

Personally I don't think we should add more broken stuff to the stdlib. In the end it's more maintenance and time lost for everyone, end-users included. Computers are frustrating to use, do not add more primitives that allow to build programs that error in weird ways.

But then I decided I would no longer spend my time trying to convince people of this project of that. So anything that I said in this discussion can be ignored.

**Comment by gasche** [comment]
> But then I decided I would no longer spend my time trying to convince people of this project of that.

Sounds like a shame! I believe that your participation on the issue tracker is appreciated, and that more often than not your recommendations are directly applied. There have been cases (recently the `input_line` stuff) where in the end the consensus went in another direction, but I hope you will continue making your opinion heard.

**Comment by novalis** [comment]
So, if @dbuenzli is out, should I put it back to the simple version?

**Comment by nojb** [comment]
> So, if @dbuenzli is out, should I put it back to the simple version?

To be honest, I am not comfortable with the lack of consensus around this PR.

I don't think we should treat the review of PRs as a kind of endurance content where the last standing person wins. Similarly, I don't think @dbuenzli being "out" is a win for any of us or the project more generally (quite the contrary, in fact).

PRs against the standad library are notoriously hard to agree on and can be frustrating for everyone involved. But my view is that we should still insist on finding consensus as far as possible. @dbuenzli has done tons of contributions to the ecosystem in general and the compiler in particular, and is a frequent and careful reviewer (a precious commodity). That makes his opinion particularly valuable to me.

Because of that, I don't feel comfortable taking a decision on this PR either way. I would be happy to pass the relay to some other maintainer to push it forward. I know this may be frustrating for @novalis (and I apologize for that and I hope you don't find it too discouraging), but the process of integrating patches from external contributors is anything but straightforward. Such inconclusive stand-offs are, sadly, rather common (and happen to everyone, maintainers included).

**Comment by gasche** [comment]
>  I would be happy to pass the relay to some other maintainer to push it forward.

Grmph. You are lucky that you have a wildly positive karma :-)

I went back to the git logs to understand the original intent behind this loop with a 1000 cutoff. The cutoff was added in https://github.com/ocaml/ocaml/commit/95933de17c122f254bd91231d04b9cab89f134c3#diff-79807ee7b626146062c834f7d687493b82bd1dc862bdd44fe96d1f3ce49101bb in 1998. Here was the resulting code:

https://github.com/ocaml/ocaml/blob/95933de17c122f254bd91231d04b9cab89f134c3/stdlib/filename.ml#L192-L205

At the time, instead of trying random file numbers (minimizing the risk of the file existing already), the code would try consecutive numbers counting upwards from 0. In case several temporary files had been already created with the same prefix and suffix, you would have to loop until finding the "next available slot". (Side note: it would be possible to avoid a quadratic behavior here by using a sort of bisection using powers of 2 to more quickly find the next available number.) Note the error message on the 1000 cutoff: "Filename.temp_file: temp dir nonexistent or full".

My understanding is that the reasoning was that there were two likely causes for errors: the directory does not exist (or, in fact, is not a directory), or there is *some* system-wide issue preventing file creation.

(This specific but also vague error message was improved into its current form in 2005 by 99885a80a0e266aa862637780d93a7b64c726d76.)

Based on this hypothetical reconstruction of the original intent:
1. I believe that it does make sense to check explicitly for the first condition, "there is a problem with the directory", and leave the other "there is a problem for file creation" for the cutoff.
2. I think that we could reduce the cutoff value a lot now that we use a PRNG to choose tempfile numbers. The PRNG is self-initialized, which means that running the same program several times will not generate the same tempfile numbers. We could move the cutoff to just 10. (The probability of seeing N conflicts is 2^(-24*N), apparently the probably of 10 failed attempts is of the order of 10^-72 -- in comparison the standard estimate for the number of atoms in the universe is 10^80).

I think that either (1) or (2) would fix the issue raised by @dbuenzli, and we could in fact probably do both.

If we decide to do (1), could we move the check (that `temp_dir` is a valid directory) instead the (module-local) `temp_file_name` function that constructs the temporary path? That is the function assuming directory structure for the `temp_dir` argument, and it would avoid duplicating the check logic three times as in the current PR.

@novalis, @sliquister, what do you think?

P.S.: the two commits from 1998 and 2005 that I dug back both have a hash that starts with 5 digits. (9/1000 chances of that happening!) Surely that cannot be a coincidence...

**Comment by xavierleroy** [comment]
This is another PR that I find not very conducive to discussions, but for what it's worth:

- I agree with @gasche's analysis about the number of rounds.  Collisions happen with 2^-12 probability (birthday paradox), so 10 collisions in a row are very unlikely indeed.  And even if the temp directory is very full, as long as it contains fewer than 2^23 files matching the expected pattern, an unused name is found with probability > 1/2 at each round, so it still converges pretty quickly (but: more than 10 rounds may be needed).

- I agree with @novalis and @sliquister that there are other failure causes than "directory not existent", e.g. "readonly file system", so I'd rather not check explicitly for directory existence.

**Comment by xavierleroy** [comment]
Two more things:

- My comments above apply both to the existing `temp_file` and the proposed `temp_dir` functions.  Personally, I don't feel a strong need for having `temp_dir` in the stdlib, but it's not absurd either.
- If we decide (against my advice) to check for directory existence, it must be done after the first failure, not before the first attempt, otherwise we're doubling the number of system calls performed, for no good reason.

**Comment by gasche** [comment]
Ah, my numbers were wrong by a square root indeed. So I guess 100 instead of 10. Thanks!

Instead of 2^24 as currently used as an upper bound for the random number, could we maybe use 2^30 to increase our confidence in the absence of conflicts?

**Comment by xavierleroy** [comment]
> So I guess 100 instead of 10. 

More like 20 instead of 10 :-)

> could we maybe use 2^30 to increase our confidence in the absence of conflicts?

We're using hexadecimal digits, so it would be either 2^28 (7 digits) or 2^32 (8 digits).  The only downside I can see is that this could break code that expects temp names to match the glob pattern `prefix??????.suffix`, e.g. shell scripts that clean up after an OCaml program.

**Comment by gasche** [comment]
Okay, 20 and no changes to the bound. (I forgot that we print in hexa.) We could also do 2^32, which requires using `Random.bits32` and `%08lx` for portability.

@novalis I guess there is a growing consensus in favor of "no directory check", which is also your preference. Would you revert to that, but also reduce the cutoff value from 1000 to 20, to improve the experience slightly in the issue-with-the-directory case?

(For the record, I think that it makes sense to check for the directory, because it provides a better error message in a non-uncommon scenario, and because this is a failure condition that is not related to the ability or not to create files in general (no space, write-only disk, out of fds etc.). But this seems to be a minority opinion.)

**Comment by novalis** [comment]
Should we also change the other cutoffs (for the other temp_*) functions, or can that be a different PR?

**Comment by gasche** [comment]
I would change all the cutoffs.

**Review by gasche — APPROVED** [review]

**Review by v-gb — APPROVED** [review]

**Comment by novalis** [comment]
Do I need more approvals?  How can we get this in?

**Comment by gasche** [comment]
I think the only thing we needed was to notice that the CI passed, but there was a conflict before that so we didn't see it and forgot about it, apologies. I rebased on the current trunk and force-pushed against your remote branch. This should be good to go when the CI agrees.

**Pull request merged**
